### PR TITLE
Value serializers

### PR DESCRIPTION
**Pull request by Dmdv**
Changes include 

- some EvaluatedValue serializers

- TwoArgumentsOperation serializers

- Relation serializers

- Tests

### CODE REVIEW

**Comment by Dmdv** [comment]
I will close this PR and create a new one to include minor fixes

File: src/main/scala/sigmastate/serialization/BigIntConstantSerializer.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    val size: Int = Ints.fromByteArray(sizeBuffer)
+
+    val valueBuffer = bytes.slice(pos + IntSize,  pos + IntSize + size)
+    BigIntConstant(new BigInteger(valueBuffer)) -> (IntSize + size)
+  }
+
+  override def serializeBody(obj: BigIntConstant): Array[Byte] = {
+
+    val data: Array[Byte] = obj.value.toByteArray
+    val length = data.length
```
Comment (kushti) [inline src/main/scala/sigmastate/serialization/BigIntConstantSerializer.scala]: let's use Short type for serialized BigInt, as we'll use big integers by some module (and module will be 521 bits max)

File: src/main/scala/sigmastate/serialization/BigIntConstantSerializer.scala [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import com.google.common.primitives.{Ints, Shorts}
+import sigmastate.SBigInt
+import sigmastate.SType.TypeCode
+import sigmastate.Values.BigIntConstant
+import sigmastate.serialization.ValueSerializer._
+
+object BigIntConstantSerializer extends ValueSerializer[BigIntConstant] {
+
+  val lengthSize: Int = 2
```
Comment (daron666) [inline src/main/scala/sigmastate/serialization/BigIntConstantSerializer.scala]: There is convention that constants should have upper camel case `LengthSize`.
[Convention Doc](https://docs.scala-lang.org/style/naming-conventions.html)

File: src/main/scala/sigmastate/serialization/ValueSerializer.scala [hunk 3]
```diff
@@ -17,45 +17,83 @@ trait ValueSerializer[V <: Value[SType]] extends SigmaSerializer[Value[SType], V
   }
 }
 
-object ValueSerializer extends SigmaSerializerCompanion[Value[SType]] {
+trait OpCodes {
```
Comment (daron666) [inline src/main/scala/sigmastate/serialization/ValueSerializer.scala]: Can we move it to a separate file?

File: src/main/scala/sigmastate/serialization/ValueSerializer.scala [hunk 4]
```diff
@@ -17,45 +17,83 @@ trait ValueSerializer[V <: Value[SType]] extends SigmaSerializer[Value[SType], V
...
+  val AppendBytesCode:    OpCode = 43: Byte
+  val ExponentiateCode:   OpCode = 44: Byte
+  val MultiplyGroupCode:  OpCode = 45: Byte
+}
+
+object ValueSerializer
+  extends SigmaSerializerCompanion[Value[SType]]
+  with OpCodes {
+
+  override type OpCode = Byte
```
Comment (daron666) [inline src/main/scala/sigmastate/serialization/ValueSerializer.scala]: Overriding `OpCode`? Why? I'm ask cause it has the same type as the inherited one - `Byte`. I suggest to remove `OpCode` from Value seialaizer, make OpCodes as an object (and move it to it's own file). Then you can eaily get this type as `OpCodes.OpCode` or `import OpCodes._`

File: src/main/scala/sigmastate/serialization/ValueSerializer.scala [hunk 5]
```diff
@@ -17,45 +17,83 @@ trait ValueSerializer[V <: Value[SType]] extends SigmaSerializer[Value[SType], V
...
     TrueLeafSerializer,
     FalseLeafSerializer,
     ConcreteCollectionSerializer,
     AndSerializer,
     OrSerializer,
-    TaggedVariableSerializer
+    TaggedVariableSerializer,
+    NotSerializer,
+    BigIntConstantSerializer,
+    ByteArrayConstantSerializer
```
Comment (kushti) [inline src/main/scala/sigmastate/serialization/ValueSerializer.scala]: Duplicate ByteArrayConstantSerializer

File: src/main/scala/sigmastate/serialization/ValueSerializer.scala [hunk 6]
```diff
@@ -17,45 +17,83 @@ trait ValueSerializer[V <: Value[SType]] extends SigmaSerializer[Value[SType], V
...
   val table = Seq[ValueSerializer[_ <: Value[SType]]](
+
     RelationSerializer(GtCode, GT.apply, Seq(Constraints.onlyInt2)),
     RelationSerializer(GeCode, GE.apply, Seq(Constraints.onlyInt2)),
     RelationSerializer(LtCode, LT.apply, Seq(Constraints.onlyInt2)),
     RelationSerializer(LeCode, LE.apply, Seq(Constraints.onlyInt2)),
     RelationSerializer(EqCode, EQ.applyNonTyped, Seq(Constraints.sameType2)),
     RelationSerializer(NeqCode, NEQ.apply, Seq(Constraints.sameType2)),
+
+    TwoArgumentsSerializerChecked(XorCode, Xor.apply),
```
Comment (kushti) [inline src/main/scala/sigmastate/serialization/ValueSerializer.scala]: Why "Checked" ?

Comment (Dmdv) [reply src/main/scala/sigmastate/serialization/ValueSerializer.scala]: There are two serializers, for this one both arguments have the same generic type.
I will think about new name

File: src/main/scala/sigmastate/serialization/ValueSerializer.scala (hunk 4 above)
Comment (Dmdv) [reply src/main/scala/sigmastate/serialization/ValueSerializer.scala]: I like the idea. Done.

File: src/main/scala/sigmastate/serialization/ValueSerializer.scala [hunk 7]
```diff
@@ -75,12 +32,12 @@ object ValueSerializer
...
 
-    TwoArgumentsSerializerChecked(XorCode, Xor.apply),
-    TwoArgumentsSerializerChecked(AppendBytesCode, AppendBytes.apply),
-    TwoArgumentsSerializer(ExponentiateCode, Exponentiate.apply, Seq(Constraints.sameType2)),
-    TwoArgumentsSerializerChecked(MultiplyGroupCode, MultiplyGroup.apply),
-    TwoArgumentsSerializerChecked(MinusCode, Minus.apply),
-    TwoArgumentsSerializerChecked(PlusCode, Plus.apply),
+    TwoArgumentsSerializer(XorCode, Xor.apply),
+    TwoArgumentsSerializer(AppendBytesCode, AppendBytes.apply),
+    TwoConstrainedArgumentsSerializer(ExponentiateCode, Exponentiate.apply, Seq(Constraints.sameType2)),
```
Comment (kushti) [inline src/main/scala/sigmastate/serialization/ValueSerializer.scala]: Why cant' TwoArgumentsSerializer and TwoConstrainedArgumentsSerializer be combined via optional constructor argument?

File: src/main/scala/sigmastate/serialization/ValueSerializer.scala (hunk 5 above)
Comment (Dmdv) [reply src/main/scala/sigmastate/serialization/ValueSerializer.scala]: Done

File: src/main/scala/sigmastate/serialization/BigIntConstantSerializer.scala [hunk 8]
```diff
@@ -0,0 +1,38 @@
+package sigmastate.serialization
+
+import java.math.BigInteger
+
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/sigmastate/serialization/BigIntConstantSerializer.scala]: @mike-aksarin we used SCAPI Java-framework for cryptography, and still using BouncyCastle (another Java framework), BigInt is introducing unnecessary converting operations then

File: src/test/scala/sigmastate/serialization/BigIntConstantSerializerSpecification.scala [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+  override def objects = Table(
+    ("object", "bytes"),
+    (BigIntConstant(BigInteger.valueOf(0)), Array[Byte](15, 0, 1, 0)),
+    (BigIntConstant(new BigInteger(Array[Byte](3,4,5))), Array[Byte](15, 1, 1, 3, 4, 5))
+  )
+
+  tableRoundTripTest("BigIntConstant: Serializer round trip")
+  tablePredefinedBytesTest("BigIntConstant: deserialize from predefined bytes")
+}
```
[the remark this answers is not in the document]
Comment (Dmdv) [reply src/test/scala/sigmastate/serialization/BigIntConstantSerializerSpecification.scala]: yes, good point

File: src/main/scala/sigmastate/serialization/ValueSerializer.scala [hunk 10]
```diff
@@ -17,45 +18,38 @@ trait ValueSerializer[V <: Value[SType]] extends SigmaSerializer[Value[SType], V
...
     RelationSerializer(GtCode, GT.apply, Seq(Constraints.onlyInt2)),
     RelationSerializer(GeCode, GE.apply, Seq(Constraints.onlyInt2)),
     RelationSerializer(LtCode, LT.apply, Seq(Constraints.onlyInt2)),
     RelationSerializer(LeCode, LE.apply, Seq(Constraints.onlyInt2)),
     RelationSerializer(EqCode, EQ.applyNonTyped, Seq(Constraints.sameType2)),
     RelationSerializer(NeqCode, NEQ.apply, Seq(Constraints.sameType2)),
+
+    TwoArgumentsSerializer(XorCode, Xor.apply),
+    TwoArgumentsSerializer(AppendBytesCode, AppendBytes.apply),
+    TwoConstrainedArgumentsSerializer(ExponentiateCode, Exponentiate.apply, Seq(Constraints.sameType2)),
```
Comment (kushti) [inline src/main/scala/sigmastate/serialization/ValueSerializer.scala]: Why same-type constraint here? Exponentiate parameters are group element and big int

**Comment by kushti** [comment]
@Dmdv please resolve conflicts and answer last questions

File: src/main/scala/sigmastate/serialization/ValueSerializer.scala (hunk 10 above)
Comment (Dmdv) [reply src/main/scala/sigmastate/serialization/ValueSerializer.scala]: I removed TwoArgumentsSerializer

File: src/test/scala/sigmastate/serialization/BigIntConstantSerializerSpecification.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (Dmdv) [reply src/test/scala/sigmastate/serialization/BigIntConstantSerializerSpecification.scala]: Added 2 more tests

**Pull request merged**